File: scripts/library-add.ts
#!/usr/bin/env tsx

import "dotenv/config";
import { loadPendingQueue, savePendingQueue } from "./library-queue-store.mjs";

type Intent = "interesting" | "try-soon" | "keep-for-later";
type SourceType = "link" | "x";
type FetchStatus = "not-needed" | "manual" | "skipped" | "failed" | "done";

type SlackRef = {
  channel: string;
  thread_ts: string;
  notified_at?: string;
};

type Draft = {
  title: string;
  summary: string;
  tags: string[];
  comment?: string;
};

type QueueItem = {
  url: string;
  note: string;
  intent: Intent | "";
  excerpt: string;
  status: "pending" | "done" | "failed";
  stage: string;
  added_at?: string;
  fetched_text: string;
  normalized_text: string;
  draft: Draft;
  error: unknown | null;
  slack?: SlackRef;
  source_type?: SourceType;
  post_id?: string;
  author_hint?: string;
  fetch_status?: FetchStatus;
  fetch_error?: string;
  fetched_at?: string;
  author?: string;
  author_name?: string;
  text?: string;
  posted_at?: string;
};

type QueueData = {
  urls?: QueueItem[];
};

type ParsedArgs = {
  url: string;
  note: string;
  intent: Intent | "";
  excerpt: string;
  slackChannel: string;
  slackThreadTs: string;
};

type XMeta = {
  sourceType: "x";
  authorHint: string;
  postId: string;
};

type XFetchResult = Partial<QueueItem> & {
  fetch_status: Exclude<FetchStatus, "not-needed" | "manual">;
};

const VALID_INTENTS = new Set<Intent>(["interesting", "try-soon", "keep-for-later"]);

function parseArgs(argv: string[]): ParsedArgs {
  const args = [...argv];
  let url = "";
  let note = "";
  let intent: Intent | "" = "";
  let excerpt = "";
  let slackChannel = "";
  let slackThreadTs = "";

  while (args.length > 0) {
    const token = args.shift();
    if (!token) continue;

    if (!url && !token.startsWith("--")) {
      url = token;
      continue;
    }

    if (token === "--note") {
      note = args.shift() || "";
      continue;
    }

    if (token === "--intent") {
      const value = args.shift() || "";
      if (value && !isIntent(value)) {
        throw new Error("intent must be one of: interesting, try-soon, keep-for-later");
      }
      intent = value ? value as Intent : "";
      continue;
    }

    if (token === "--excerpt") {
      excerpt = args.shift() || "";
      continue;
    }

    if (token === "--slack-channel") {
      slackChannel = args.shift() || "";
      continue;
    }

    if (token === "--slack-thread-ts") {
      slackThreadTs = args.shift() || "";
      continue;
    }

    throw new Error(`unknown argument: ${token}`);
  }

  if (!url) {
    throw new Error("usage: ./library.sh add <url> [--note <text>] [--intent <value>] [--excerpt <text>]");
  }

  return { url, note, intent, excerpt, slackChannel, slackThreadTs };
}

function isIntent(value: string): value is Intent {
  return VALID_INTENTS.has(value as Intent);
}

function now(): string {
  return new Date().toISOString();
}

function updateExistingItem(
  items: QueueItem[],
  url: string,
  patch: Partial<Pick<QueueItem, "note" | "intent" | "excerpt" | "slack">>,
): QueueItem | null {
  const item = items.find((entry) => entry.url === url && entry.status === "pending");
  if (!item) return null;

  if (patch.slack) {
    item.slack = {
      ...(item.slack || {}),
      ...patch.slack,
    };
  }
  if (patch.note !== undefined && patch.note !== "") item.note = patch.note;
  if (patch.intent !== undefined && patch.intent !== "") item.intent = patch.intent;
  if (patch.excerpt !== undefined && patch.excerpt !== "") item.excerpt = patch.excerpt;

  if (item.source_type === "x" && patch.excerpt) {
    item.fetch_status = "manual";
    delete item.fetch_error;
  }

  item.stage = item.stage || "queued";
  item.error = null;

  return item;
}

function extractXMeta(rawUrl: string): XMeta | null {
  let parsed: URL;
  try {
    parsed = new URL(rawUrl);
  } catch {
    return null;
  }

  if (!["x.com", "www.x.com", "twitter.com", "www.twitter.com"].includes(parsed.hostname)) {
    return null;
  }

  const match = parsed.pathname.match(/^\/([^/]+)\/status\/(\d+)/);
  if (!match) {
    return null;
  }

  return { sourceType: "x", authorHint: match[1], postId: match[2] };
}

async function fetchXPost(postId: string): Promise<XFetchResult> {
  const token = process.env.X_BEARER_TOKEN;
  if (!token) {
    return { fetch_status: "skipped", fetch_error: "X_BEARER_TOKEN is not set" };
  }

  const endpoint = new URL(`https://api.x.com/2/tweets/${postId}`);
  endpoint.searchParams.set("expansions", "author_id");
  endpoint.searchParams.set("tweet.fields", "author_id,created_at,text");
  endpoint.searchParams.set("user.fields", "username,name");

  const response = await fetch(endpoint, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });

  const payload = await response.json().catch(() => ({})) as {
    detail?: string;
    title?: string;
    data?: { text?: string; created_at?: string };
    includes?: { users?: Array<{ username?: string; name?: string }> };
  };

  if (!response.ok) {
    const message = payload.detail || payload.title || `HTTP ${response.status}`;
    return { fetch_status: "failed", fetch_error: message };
  }

  const author = payload.includes?.users?.[0] || {};
  return {
    fetch_status: "done",
    fetched_at: now(),
    author: author.username || "",
    author_name: author.name || "",
    text: payload.data?.text || "",
    posted_at: payload.data?.created_at || "",
  };
}

async function main(): Promise<void> {
  const { url, note, intent, excerpt, slackChannel, slackThreadTs } = parseArgs(process.argv.slice(2));
  const queue = loadPendingQueue() as QueueData;
  const items = Array.isArray(queue.urls) ? queue.urls : [];

  const duplicate = items.find((item) => item.url === url && item.status === "pending");
  if (duplicate) {
    const updated = updateExistingItem(items, url, {
      note,
      intent,
      excerpt,
      slack: slackChannel ? { channel: slackChannel, thread_ts: slackThreadTs || "" } : undefined,
    });
    savePendingQueue(queue);

    const lines = [`⚠ すでにカウンターにある: ${url}`];
    if (updated?.excerpt) lines.push("  抜粋: 更新");
    if (updated?.note) lines.push(`  ひとこと: ${updated.note}`);
    if (updated?.intent) lines.push(`  意図: ${updated.intent}`);
    if (updated?.fetch_status) lines.push(`  状態: ${updated.fetch_status}`);
    console.log(lines.join("\n"));
    return;
  }

  const item: QueueItem = {
    url,
    note,
    intent,
    excerpt,
    status: "pending",
    stage: "queued",
    added_at: now(),
    fetched_text: "",
    normalized_text: "",
    draft: {
      title: "",
      summary: "",
      tags: [],
      comment: "",
    },
    error: null,
  };

  if (slackChannel) {
    item.slack = {
      channel: slackChannel,
      thread_ts: slackThreadTs || "",
      notified_at: "",
    };
  }

  const xMeta = extractXMeta(url);
  if (xMeta) {
    item.source_type = xMeta.sourceType;
    item.post_id = xMeta.postId;
    item.author_hint = xMeta.authorHint;

    if (excerpt) {
      item.fetch_status = "manual";
    } else {
      Object.assign(item, await fetchXPost(xMeta.postId));
    }
  } else {
    item.source_type = "link";
    item.fetch_status = "not-needed";
  }

  items.push(item);
  queue.urls = items;
  savePendingQueue(queue);

  const lines = [
    `✓ 図書室カウンターに追加: ${url}`,
    item.source_type === "x"
      ? `  X入力: ${item.fetch_status}${item.author ? ` / @${item.author}` : ""}`
      : `  種別: ${item.source_type}`,
  ];

  if (item.fetch_error) {
    lines.push(`  理由: ${item.fetch_error}`);
  }
  if (excerpt) {
    lines.push("  抜粋: あり");
  }
  if (intent) {
    lines.push(`  意図: ${intent}`);
  }
  if (note) {
    lines.push(`  ひとこと: ${note}`);
  }

  console.log(lines.join("\n"));
}

main().catch((error: unknown) => {
  const message = error instanceof Error ? error.message : String(error);
  console.error(`✗ ${message}`);
  process.exit(1);
});

File: scripts/library-add.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "json"
require "net/http"
require "uri"

$LOAD_PATH.unshift(File.expand_path("../lib", __dir__))
require "usurahi/library_queue_store"

VALID_INTENTS = %w[interesting try-soon keep-for-later].freeze

def load_dotenv
  env_path = File.expand_path("../.env", __dir__)
  return unless File.file?(env_path)

  File.each_line(env_path) do |line|
    stripped = line.strip
    next if stripped.empty? || stripped.start_with?("#") || !stripped.include?("=")

    key, value = stripped.split("=", 2)
    ENV[key] ||= value.to_s.gsub(/\A["']|["']\z/, "")
  end
end

def parse_args(argv)
  args = argv.dup
  parsed = {
    "url" => "",
    "note" => "",
    "intent" => "",
    "excerpt" => "",
    "slack_channel" => "",
    "slack_thread_ts" => ""
  }

  until args.empty?
    token = args.shift
    if parsed["url"].empty? && !token.start_with?("--")
      parsed["url"] = token
      next
    end

    case token
    when "--note"
      parsed["note"] = args.shift.to_s
    when "--intent"
      parsed["intent"] = args.shift.to_s
    when "--excerpt"
      parsed["excerpt"] = args.shift.to_s
    when "--slack-channel"
      parsed["slack_channel"] = args.shift.to_s
    when "--slack-thread-ts"
      parsed["slack_thread_ts"] = args.shift.to_s
    else
      raise ArgumentError, "unknown argument: #{token}"
    end
  end

  if parsed["url"].empty?
    raise ArgumentError, "usage: ./library.sh add <url> [--note <text>] [--intent <value>] [--excerpt <text>]"
  end

  if !parsed["intent"].empty? && !VALID_INTENTS.include?(parsed["intent"])
    raise ArgumentError, "intent must be one of: interesting, try-soon, keep-for-later"
  end

  parsed
end

def now
  Time.now.utc.iso8601
end

def update_existing_item(items, url, patch)
  item = items.find { |entry| entry["url"] == url && entry["status"] == "pending" }
  return nil unless item

  patch.each do |key, value|
    if key == "slack" && value.is_a?(Hash)
      item["slack"] = (item["slack"] || {}).merge(value)
      next
    end
    item[key] = value unless value == ""
  end

  if item["source_type"] == "x" && patch["excerpt"].to_s != ""
    item["fetch_status"] = "manual"
    item.delete("fetch_error")
  end

  item["stage"] ||= "queued"
  item["error"] = nil
  item
end

def extract_x_meta(raw_url)
  parsed = URI.parse(raw_url)
  return nil unless %w[x.com www.x.com twitter.com www.twitter.com].include?(parsed.host)

  match = parsed.path.match(%r{\A/([^/]+)/status/(\d+)})
  return nil unless match

  { "source_type" => "x", "author_hint" => match[1], "post_id" => match[2] }
rescue URI::InvalidURIError
  nil
end

def fetch_x_post(post_id)
  token = ENV["X_BEARER_TOKEN"]
  return { "fetch_status" => "skipped", "fetch_error" => "X_BEARER_TOKEN is not set" } if token.to_s.empty?

  endpoint = URI("https://api.x.com/2/tweets/#{post_id}")
  params = {
    "expansions" => "author_id",
    "tweet.fields" => "author_id,created_at,text",
    "user.fields" => "username,name"
  }
  endpoint.query = URI.encode_www_form(params)

  request = Net::HTTP::Get.new(endpoint)
  request["Authorization"] = "Bearer #{token}"
  response = Net::HTTP.start(endpoint.hostname, endpoint.port, use_ssl: true) { |http| http.request(request) }
  payload = begin
    JSON.parse(response.body)
  rescue JSON::ParserError
    {}
  end
  normalize_x_response(response, payload)
rescue JSON::ParserError
  { "fetch_status" => "failed", "fetch_error" => "invalid JSON response" }
rescue StandardError => e
  { "fetch_status" => "failed", "fetch_error" => e.message }
end

def normalize_x_response(response, payload)
  unless response.is_a?(Net::HTTPSuccess)
    message = payload.fetch("detail", nil) || payload.fetch("title", nil) || "HTTP #{response.code}"
    return { "fetch_status" => "failed", "fetch_error" => message }
  end

  author = payload.dig("includes", "users", 0) || {}
  {
    "fetch_status" => "done",
    "fetched_at" => now,
    "author" => author["username"] || "",
    "author_name" => author["name"] || "",
    "text" => payload.dig("data", "text") || "",
    "posted_at" => payload.dig("data", "created_at") || ""
  }
end

def main(argv)
  load_dotenv
  args = parse_args(argv)
  queue = Usurahi::LibraryQueueStore.load_pending_queue
  items = queue["urls"].is_a?(Array) ? queue["urls"] : []

  if items.any? { |item| item["url"] == args["url"] && item["status"] == "pending" }
    updated = update_existing_item(items, args["url"], {
      "note" => args["note"],
      "intent" => args["intent"],
      "excerpt" => args["excerpt"],
      "slack" => args["slack_channel"].empty? ? nil : {
        "channel" => args["slack_channel"],
        "thread_ts" => args["slack_thread_ts"]
      }
    })
    Usurahi::LibraryQueueStore.save_pending_queue(queue)

    lines = ["⚠ すでにカウンターにある: #{args["url"]}"]
    lines << "  抜粋: 更新" unless updated["excerpt"].to_s.empty?
    lines << "  ひとこと: #{updated["note"]}" unless updated["note"].to_s.empty?
    lines << "  意図: #{updated["intent"]}" unless updated["intent"].to_s.empty?
    lines << "  状態: #{updated["fetch_status"]}" unless updated["fetch_status"].to_s.empty?
    puts lines.join("\n")
    return
  end

  item = {
    "url" => args["url"],
    "note" => args["note"],
    "intent" => args["intent"],
    "excerpt" => args["excerpt"],
    "status" => "pending",
    "stage" => "queued",
    "added_at" => now,
    "fetched_text" => "",
    "normalized_text" => "",
    "draft" => {
      "title" => "",
      "summary" => "",
      "tags" => [],
      "comment" => ""
    },
    "error" => nil
  }

  unless args["slack_channel"].empty?
    item["slack"] = {
      "channel" => args["slack_channel"],
      "thread_ts" => args["slack_thread_ts"],
      "notified_at" => ""
    }
  end

  x_meta = extract_x_meta(args["url"])
  if x_meta
    item["source_type"] = x_meta["source_type"]
    item["post_id"] = x_meta["post_id"]
    item["author_hint"] = x_meta["author_hint"]

    if args["excerpt"].empty?
      item.merge!(fetch_x_post(x_meta["post_id"]))
    else
      item["fetch_status"] = "manual"
    end
  else
    item["source_type"] = "link"
    item["fetch_status"] = "not-needed"
  end

  items << item
  queue["urls"] = items
  Usurahi::LibraryQueueStore.save_pending_queue(queue)

  lines = [
    "✓ 図書室カウンターに追加: #{args["url"]}",
    if item["source_type"] == "x"
      "  X入力: #{item["fetch_status"]}#{item["author"].to_s.empty? ? "" : " / @#{item["author"]}"}"
    else
      "  種別: #{item["source_type"]}"
    end
  ]
  lines << "  理由: #{item["fetch_error"]}" unless item["fetch_error"].to_s.empty?
  lines << "  抜粋: あり" unless args["excerpt"].empty?
  lines << "  意図: #{args["intent"]}" unless args["intent"].empty?
  lines << "  ひとこと: #{args["note"]}" unless args["note"].empty?
  puts lines.join("\n")
end

begin
  main(ARGV)
rescue StandardError => e
  warn "✗ #{e.message}"
  exit 1
end
